paddle/fluid/distributed/auto_parallel/spmd_rules/elementwise_spmd_rule.cc

```diff
@@ -0,0 +1,137 @@
+/* Copyright (c) 2023 PaddlePaddle Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#include "paddle/fluid/distributed/auto_parallel/spmd_rules/elementwise_spmd_rule.h"
+#include "paddle/phi/core/distributed/auto_parallel/utils.h"
+
+namespace paddle {
+namespace distributed {
+namespace auto_parallel {
+using phi::distributed::auto_parallel::str_join;
+
+std::pair<std::vector<TensorDistAttr>, std::vector<TensorDistAttr>>
+ElementwiseSPMDRule::InferForward(
+    const std::vector<DistTensorSpec>& input_specs,
+    const paddle::framework::AttributeMap& attrs) {
+  // step0: Verify Input Args Based on Elementwise Logic
+  int64_t ninputs = input_specs.size();
+  PADDLE_ENFORCE_GT(
+      ninputs,
+      0,
+      phi::errors::InvalidArgument("The size of InputSpec in elementwise must "
+                                   "be greater than 0, but got [%d].",
+                                   ninputs));
+  VerifySpecs(input_specs, "elementwise");
+
+  // step1: Build Einsum Notation
+  std::string alphabet = "abcdefghijklmnopqrstuvwxyz";
+  std::vector<std::string> input_axes_vec;
+  int64_t max_ndim = 0;
+  for (int64_t i = 0; i < ninputs; ++i) {
+    int64_t ndim = input_specs[i].shape().size();
+    if (ndim > max_ndim) {
+      max_ndim = ndim;
+    }
+  }
+
+  // get einsum notation for each input, deal with broadcast
+  std::vector<int64_t> broadcast_axis_count(max_ndim, 0);
+  for (int64_t i = 0; i < ninputs; ++i) {
+    std::vector<int64_t> shape = input_specs[i].shape();
+    int64_t ndim = shape.size();
+    int64_t start_dim = max_ndim - ndim;
+    std::string axes_notation = GetBroadcastAxes(ndim, max_ndim, alphabet);
+    if (ninputs > 1) {
+      for (int64_t idim = 0; idim < max_ndim; idim++) {
+        // deal with the broadcast axes, record the
+        // input number at each broadcast axis
+        if (idim < start_dim) {
+          broadcast_axis_count[idim] += 1;
+        } else if (shape[idim - start_dim] == 1) {
+          broadcast_axis_count[idim] += 1;
+          // mark the broadcast axis to a special "1"
+          axes_notation[idim - start_dim] = '1';
```

Comment:
"1" concept is not need for sharding merge.

since we assume that the income mapping is correct for spmd rule, if any tensor axis's size is "1", and the dim_mapping for this axis should be "-1" correspondingly.  and if all inputs' dim_mappings are "-1", it merge to "-1" with no doubt.

but it is ok here.

Reply:
Here "1" is a special label for broadcasting dim, with this label broadcasting case can handled with the same function as common cases.